Updates to ISY994 Component (Climate 1.0, Improved Z-Wave Device Support, Fix Attribute and Heartbeat Issues)

shbatm opened the pull request:
## Description:
This PR is to update the ISY994 component with various bug fixes, better Z-Wave integration, device state attribute support, ISY Variables support, and add Climate support for Insteon and Z-Wave Thermostats that are exposed by the ISY994 Controller hub component.

Change Log:

- Move constants to a dedicated `const.py` file.
- Add support for Climate devices per `climate-1.0` changes.
- Expand ISY Z-Wave support by properly classifying devices using the `devtype.cat` attribute.
- Fixes issues with `device_state_attributes` throwing Type errors and getting overwritten on updates from the ISY--now maintains a updatable dict and adds attributes when provided from the ISY EventStream (ISY doesn't always provide all attributes on the initial query).
- Fixes ISY Heartbeats per #21996
- Update ISYBinarySensorHeartbeat to report state on startup if available.
- Revised comments to better clarify the behaviors of the heartbeat node in Home Assistant since it does not report the actual state of the ISY's sub-node (which could be on or off).
- Add logic for Motion Sensor subnodes, update framework for Z-Wave IDs
- Allow ISY Variables to be used as sensors, binary_sensors, and switches.
- The `isy994_control` event has been updated to also expose `value`.  This is expected to be a non-breaking change in PyISY; (PyISY's `event` is now a `dict` instead of a `string`, but will represent itself the same as the command string it used to pass).  This change is vital to the `climate` integration, as some values are only made available from the event stream, not the node definition.


### HOLD PENDING PACKAGE UPDATE
Please note, this PR requires an update to be included in the PyISY package.  Currently, it is pending https://github.com/automicus/PyISY/pull/44 which fixes some malformed commands for controlling the devices.  I am submitting this here now to make it available for review and make any changes required while the package changes are pending.  Changes can be tested with the PyISY_beta package, but the intent is to include those changes in the original PyISY package.

**Related issue (if applicable):** Also includes fix for #21996

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#8475

## Example entry for `configuration.yaml` (if applicable):
```yaml
isy994:
  host: !secret isy_url
  username: !secret isy_username
  password: !secret isy_password
  isy_variables:
    sensors:
      - id: 23
        type: 2
        unit_of_measurement: '%'
        device_class: 'battery'
    binary_sensors:
      - id: 14
        type: 2
    switches:
      - id: 5
        type: 2
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [n/a] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [n/a] Tests have been added to verify that the new code works.
  - Tested and functioning with my ISY device.

balloob commented on homeassistant/components/climate/isy994.py:
```diff
@@ -0,0 +1,321 @@
+"""
```
Please rename this to `isy994/climate.py`

balloob commented on homeassistant/components/climate/isy994.py:
```diff
@@ -0,0 +1,321 @@
+Support for Insteon Thermostats via ISY994 Platform.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/light.isy994/
```
Remove this paragraph

balloob commented on homeassistant/components/climate/isy994.py:
```diff
@@ -0,0 +1,321 @@
+DEFAULT_MAX_TEMP = 30
+
+# Modes not currently supported by 'climate' domain, but may be sent from ISY.
+STATE_PROG_HEAT = 'program_heat'
```
You are not allowed to make up states. You need to follow the states of the climate component.

shbatm replied:
Updated to still track the operation mode as reported by the ISY, but to convert these modes into the correct Climate State and hold modes.  Hold modes appear to only be partially implemented at this time (no list of approved hold mode states--e.g. program & temp).

balloob commented on homeassistant/components/climate/isy994.py:
```diff
@@ -0,0 +1,321 @@
+        self._uom = ''
+        if len(self._node.uom) == 1:
+            self._uom = self._node.uom[0]
+        self._current_status = STATE_UNKNOWN
```
Do not use `STATE_UNKNOWN`, instead use `None`

balloob commented on homeassistant/components/climate/isy994.py:
```diff
@@ -0,0 +1,321 @@
+        return [STATE_HEAT,
+                STATE_COOL,
+                STATE_AUTO,
+                STATE_PROG_AUTO,
```
Remove

balloob commented on homeassistant/components/climate/isy994.py:
```diff
@@ -0,0 +1,321 @@
+    def device_state_attributes(self):
+        """Return the device specific state attributes."""
+        data = {
+            ATTR_CURRENT_TEMPERATURE: self.current_temperature,
```
Don't list all these things again, they are already handled by the climate class.

shbatm replied:
I attempted to remove this and got several errors because when `device_state_attributes` is called it falls back to the definition in `isy994/__init__.py` which doesn't handle the response correctly.

balloob commented on homeassistant/components/climate/isy994.py:
```diff
@@ -0,0 +1,321 @@
+                self._target_temp_low:
+            self._node.climate_setpoint_heat(int(kwargs['target_temp_low']))
+            self._target_temp_low = kwargs['target_temp_low']
+        self.schedule_update_ha_state()
```
These are not needed, this is a polling device so HA will already do this after calling this method.

shbatm replied:
ISY is a Local Push and the states on the front end did not update correctly when I removed these lines.

balloob commented on homeassistant/components/isy994/__init__.py:
```diff
@@ -17,7 +17,8 @@
 
-REQUIREMENTS = ['PyISY==1.1.1']
+REQUIREMENTS = [
+    'https://github.com/shbatm/PyISY/archive/develop.zip#PyISY==1.1.2']
```
Only PyPi published packages are allowed

shbatm replied:
Just an update, I am still working on incorporating the changes requested above; however, I am also still waiting on the upstream package to be updated--this requirement will be reverted to the PyPi package once updated.  See note in original pull request description.

MartinHjelmare commented on homeassistant/components/isy994/climate.py:
```diff
@@ -0,0 +1,360 @@
+                VALUE_TO_HASS_MODE.get(current_operation['value'], None)
+
+        await super().async_added_to_hass()
+        self.async_schedule_update_ha_state()
```
Remove this. It will be called as the last step of the entity addition.

MartinHjelmare commented on homeassistant/components/isy994/climate.py:
```diff
@@ -0,0 +1,360 @@
+        self.schedule_update_ha_state()
+
+    @property
+    def state(self):
```
Don't overwrite this. It's the base climate entity's responsibility. Implement `current_operation` and/or `is_on` as needed.

shbatm replied:
For this I'd like to see some resolution to home-assistant/architecture#22 before final acceptance.  In my opinion, current state ≠ current operation.  The base climate entity currently uses `current_operation` as it's state, but that creates mis-matched information:  I can have my current_operation mode set to `HEAT`, but the furnace might not be on at the moment because it is within temperature range--if I am reporting `state` as `current_operation` I have no way of knowing if the unit is actually running or not.

MartinHjelmare replied:
There's no timeline for arch issue 22. I would suggest you follow my change request so we can get this PR merged. When arch issue 22 is resolved we can update the integration.

MartinHjelmare commented on homeassistant/components/isy994/climate.py:
```diff
@@ -0,0 +1,360 @@
+    @property
+    def operation_list(self):
+        """Return the list of available operation modes."""
+        return [STATE_HEAT,
```
Extract this constant list to a module level constant.

MartinHjelmare commented on homeassistant/components/isy994/climate.py:
```diff
@@ -0,0 +1,360 @@
+    @property
+    def supported_features(self):
+        """Return the list of supported features."""
+        return (SUPPORT_OPERATION_MODE |
```
Extract this into a module level constant.

MartinHjelmare commented on homeassistant/components/isy994/climate.py:
```diff
@@ -0,0 +1,360 @@
+    @property
+    def temperature_unit(self):
+        """Return the unit of measurement."""
+        if self._temp_unit:
```
How can we report current and target temperatures without knowing the unit of those values?

shbatm replied:
I believe this is a limitation of the ISY and how it handles climate devices--not all devices report back a unit and those that do don't report it on initial query, it only comes through on the event log after connection.  In the code I assume the user is using the same units in HA as the ISY, until the ISY actually passes the UOM.

MartinHjelmare commented on homeassistant/components/isy994/climate.py:
```diff
@@ -0,0 +1,360 @@
+    def device_state_attributes(self):
+        """Return the device specific state attributes."""
+        data = {
+            ATTR_CURRENT_TEMPERATURE: self.current_temperature,
```
Don't report the same things as the climate entity will report. Just return `None` here if we want to block the `ISYDevice` property.

MartinHjelmare commented on homeassistant/components/isy994/climate.py:
```diff
@@ -0,0 +1,360 @@
+}
+
+FAN_MODE_TO_VALUE = {}
+for key in FAN_VALUE_TO_MODE:
```
```suggestion
FAN_MODE_TO_VALUE = {val: key for key, val in FAN_VALUE_TO_MODE.items()}
```

MartinHjelmare commented on homeassistant/components/isy994/climate.py:
```diff
@@ -0,0 +1,360 @@
+}
+
+MODE_TO_VALUE = {}
+for key in VALUE_TO_ISY_MODE:
```
```suggestion
MODE_TO_VALUE = {val: key for key, val in VALUE_TO_ISY_MODE.items()}
```

MartinHjelmare commented on homeassistant/components/isy994/climate.py:
```diff
@@ -0,0 +1,360 @@
+        self._target_temp_high = 0
+        self._fan_mode = None
+        self._current_operation = None
+        self._current_isy_operation = None
```
This looks like it's only assigned, not used.

shbatm replied:
It's set to the mode that is reported back from the ISY in Line 142 which I am splitting out into `operation_mode` and `hold_mode` to meet the climate modules' requirements for mode.  It was left in there for debugging and as a placeholder depending on the direction home-assistant/architecture#22 goes.

MartinHjelmare replied:
Unused attributes should be removed before merge.

MartinHjelmare commented on homeassistant/components/isy994/climate.py:
```diff
@@ -0,0 +1,345 @@
+        """Set new target fan mode."""
+        _LOGGER.debug('Requested fan mode %s', fan_mode)
+        # pylint: disable=protected-access
+        self._node._fan_mode(fan_mode)
```
Why do we need to access private methods? It would be best to either use the intended way of changing mode or to change the library to accessing a private method.

MartinHjelmare commented on homeassistant/components/isy994/manifest.json:
```diff
@@ -3,7 +3,7 @@
   "documentation": "https://www.home-assistant.io/components/isy994",
   "requirements": [
-    "PyISY==1.1.1"
+    "https://github.com/shbatm/PyISY/archive/develop.zip#PyISY==1.1.2"
```
This needs to be fixed before merge.

shbatm replied:
Agreed. See comment in original PR description.  I am not the owner of PyISY and have been trying to give him the time to review and accept the PR over there. This cannot be accepted until that change is integrated.  I will give it 1 more week before I push a forked version to PyPi.

shbatm replied:
I updated the lines for now and split off my testing version to another branch.

##PLEASE DO NOT MERGE UNITL PyISY HAS BEEN UDPATED

MartinHjelmare commented on requirements_all.txt:
```diff
@@ -45,7 +45,7 @@ OPi.GPIO==0.3.6
 
 # homeassistant.components.isy994
-PyISY==1.1.1
+https://github.com/shbatm/PyISY/archive/develop.zip#PyISY==1.1.2
```
See above.

shbatm commented:
### NOTE: PLEASE DO NOT MERGE UNTIL PyISY HAS BEEN UPDATED TO 1.1.2

MartinHjelmare commented:
@shbatm ping me when the library is updated.

shbatm commented:
Updated the PyPi requirements to use PyISY_beta==1.1.2.  This is a fork of the original PyISY I have pushed to PyPi.  This was done because I have not received any response from the PyISY developer in several months.  I'm not really in favor of accepting this PR using a forked requirement like this, but I will leave it up to the reviewers if they want to accept it for now and if the original PyISY is updated, reset the dependency.

MartinHjelmare commented on homeassistant/components/isy994/binary_sensor.py:
```diff
@@ -81,6 +85,15 @@ def _detect_device_type(node) -> str:
+                # pylint: disable=no-else-return
+                if subnode_id == 3:
+                    return 'heat'
+                else:
```
Remove the `else:` and we can outdent the `return 'cold'` below. Then we can remove the pylint disable.

shbatm replied:
Corrected.  Sorry, I thought that had been removed. May have lost the change during the rebase.

MartinHjelmare commented:
It would be good if there is a github fork announced together with the pypi fork, so that other community members that want to continue evolving PyISY can rally together on a common repo.

MartinHjelmare commented:
I see there was some activity from the maintainer in one PR about a month ago in the PyISY repo. On the other hand there are PRs which haven't had a reply for years. 🤷‍♂️

shbatm commented:
Yeah, it's been hit or miss with previous changes and I know there's a lot of active development on a different ISY interface that uses the library so I've been hesitant to run off in my own direction.

MartinHjelmare commented:
@balloob what do you want to do? Should we change to the pypi PyISY-beta fork?

balloob commented:
I know that maintainer. Texted him to see if he prefers we fork or we can add people as collaborators.

rmkraus commented:
Yeah, that maintainer guy is the worst. 😉 

I gave Paulus owner access to the PyISY repo and the PyPI project so he can help with keeping it up to date. I highly suggest that someone else take over development of the PyISY library. I actually don't have any ISY's connected at the moment so I don't really have any way of maintaining the integrity of the library.

balloob commented:
Alright, I will be giving @OverloadUT access to PyISY, and he will take it from here.

OverloadUT commented:
Confirming that I am taking over maintainance of the PyISY library. I'll start going over the PRs on that side, starting with the one that this change depends on!

shbatm commented:
@rmkraus -- Thank you for the development and support thus far, awesome module!

@balloob / @OverloadUT -- Thank you both as well for the support on this. Just want to let you know I'll be away for the next ~10 days.  Anything you need from me please let me know and I will address when I return.  There have been a couple people [from the community](https://community.home-assistant.io/t/insteon-thermostat-2441th-for-isy/97821/23) testing this change as a `custom_component` and I know there is one issue that has not yet been investigated with a Z-Wave Thermostat, raised at (shbatm/home-assistant#1), and for @OverloadUT: [my PyISY_beta branch](https://github.com/shbatm/PyISY/tree/PyISY_beta) has automicus/PyISY#46 incorporated as well as my changes from automicus/PyISY#44.

shbatm commented:
Just an update: I'm watching #23899 and once there is some resolution and consensus with things being discussed there, I will move forward with updating this PR to match the new climate 1.0 framework.  For the time being I have pushed a re-base to the current `dev` branch for my testing.  Please disregard and continue to hold this where it is.

OverloadUT commented:
Thanks for the update. I'll be sure to get the PyISY changes merged to align with those changes

shbatm committed:
Add Insteon Thermostat support to ISY994 Component

shbatm commented:
Original PR has been amended to include a broader scope of changes to the ISY994 platform.  These are being included as one PR because the improvements have become intertwined with the addition of the climate platform.

Relevant feedback from the forums, including testers:
https://community.home-assistant.io/t/insteon-thermostat-2441th-for-isy
https://community.home-assistant.io/t/isy994-z-wave-sensor-enhancements-testers-wanted
https://community.home-assistant.io/t/isy994-variables-as-sensors
https://community.home-assistant.io/t/isy994-zwave-multilevel-sensor-6

shbatm committed:
ISY Z-Wave Updates, Heartbeat Fix, and Move Constants to Separate File

Fixes ISY Heartbeats per home-assistant/home-assistant#21996

Update ISYBinarySensorHeartbeat to report state on startup if available.

Revised comments to better clarify the behaviors of the heartbeat node in Home Assistant since it does not report the actual state of the ISY's sub-node (which could be on or off).

Add logic for Motion Sensor subnodes, update framework for Z-Wave IDs

Allow ISY Variables to be used as sensors, binary_sensors, and switches.

Adhere to developer standards for state attributes
Fix merge error in returning attributes.


Update Requirements for PyISY to PyISY_beta==1.1.10


Correct constants for ISY Variables schema

Remove useless super().__init__ call per PyLint.


Fix wrong constant used for ON/OFF Variable Config Values

shbatm committed:
Changes for Climate-1.0 Updates


Remove errant error log statement.


Sort imports for ISY Climate


Update PyISY_beta to 1.1.11


Fix typo--setting target_temp_high with low value.


Fix casting issues in initial climate device setup.

shbatm committed:
Use async methods for loading; wait for Hass Start to get events

shbatm committed:
Add Motion Sensor II as `motion` device.


Updates for better Motion Sensor II support. Fix mis-categorizations.


Bump PyISY_beta version to 1.1.12 - Bugfix for Node with no state

shbatm committed:
Bump PyISY-beta to 1.1.13

shbatm committed:
Fix binary_sensor/sensor status not updating. Bump PyISY_beta to 1.1.14


Fix sensor not updating.

MartinHjelmare commented:
Please consider splitting this PR into multiple logical parts. It's already too big.

shbatm committed:
Update sensors to correctly check UOM for both ISYv4 and ISYv5

powersra commented:
Thanks so much for working on this, shbatm!  Great job and much appreciated.

shbatm commented:
Closing this PR for now until PyISY changes are implemented. All changes are available on my `isy994-climate` fork and can be tested as a custom component. Discussion will continue in the forum links above.

Once PyISY changes are implemented I will resubmit several PRs for the various changes implemented here.
